Source/Protected.swift

```diff
@@ -49,35 +49,27 @@ extension Lock {
     }
 }
 
-#if os(Linux)
-/// A `pthread_mutex_t` wrapper.
-final class MutexLock: Lock {
-    private var mutex: UnsafeMutablePointer<pthread_mutex_t>
-
-    init() {
-        mutex = .allocate(capacity: 1)
-
-        var attr = pthread_mutexattr_t()
-        pthread_mutexattr_init(&attr)
-        pthread_mutexattr_settype(&attr, .init(PTHREAD_MUTEX_ERRORCHECK))
+#if os(Linux) || os(Windows)
 
-        let error = pthread_mutex_init(mutex, &attr)
-        precondition(error == 0, "Failed to create pthread_mutex")
-    }
+/// An `NSLock` wrapper.
+final class MutexLock: Lock {
+    private var mutex = NSLock()
+    private var locked = false
```

Comment:
Do we need this at all? Shouldn't `NSLock` check this itself?

Reply:
I just left it so if our wrapper goes out of scope and gets deallocated, the mutex still unlocks. To be honest, now that I’m thinking about it. We should use `DispatchQueue` instead of NSLock. NSLock has the following disadvantage:

```
The NSLock class uses POSIX threads to implement its locking behavior. When sending an unlock message to an NSLock object, you must be sure that message is sent from the same thread that sent the initial lock message. Unlocking a lock from a different thread can result in undefined behavior.
```

Reply:
We already make that guarantee on Alamofire's internal queues with our closure wrapper. `DispatchQueue`s are basically the slowest locking implementation and so we avoid them for simply guaranteeing state safety.

In any case, I don't think we need to track this locking at all, just let the lock do it if it's important.

Reply:
Got it

Reply:
Pushed